#!/usr/bin/env Rscript

## The purpose of this script is
## 2. differential expression analysis (DEA) of MLL2 KO/WT cell lines, using DESeq2 (after featureCounts)
## 3. brief visualization of DEA results


## Boiler plate
library(DESeq2)
library(ggplot2)
library(reshape2)
library(gplots)
library(gridExtra)
library(pheatmap)
library(grid)


##------------------------------------------------------------------------------------------------------------------------
##  DATA
##------------------------------------------------------------------------------------------------------------------------

#custom palettes
source('/projects/jtopham_prj/Thesis_prj/scripts/custom_palettes.R')

# COSMIC cancer gene census (http://cancer.sanger.ac.uk/census/)
cosmic.genes<-read.delim("/projects/jtopham_prj/Thesis_prj/data/HEK_RNAseq/filtered_cosmic_consensus.tsv",sep='\t',header=F,stringsAsFactors=F)

# chromosome-indexed list of hg19 genes
hg19.genes<-read.delim("/projects/jtopham_prj/Thesis_prj/data/HEK_RNAseq/chr_indexed_genelist.tsv",sep='\t',header=F,stringsAsFactors=F)

# mapping file from ENSEMBL (human to mouse ortholog)
mapper<-read.delim('//projects/jtopham_prj/Thesis_prj/data/souroullas_et_al/mart_export.txt',sep='\t',header=T,stringsAsFactors=F)

# list of DE genes from Ortega et al (2015) with pval < 0.05
# (Genes downregulated in B220 positive lymphoma cells from VavPBcl2 sh-Kmt2d tumors vs VavPBcl2-vector tumors)
ortega<-read.delim('/projects/jtopham_prj/Thesis_prj/data/HEK_RNAseq/ortega_mouse_DE_down.txt',header=F,sep='\t',stringsAsFactors=F)
ortega<-unique(mapper$Associated.Gene.Name[mapper$Mouse.associated.gene.name%in%ortega[,1]])

# mapping file for ensembl ids to gene ids
ensembl.mapper<-read.delim("/projects/jtopham_prj/Thesis_prj/data/HEK_RNAseq/ensembl_mapper.tsv",sep='\t',header=T,stringsAsFactors=F)

# gene annotations from ensembl file (to annotate gene ids with 'biotype')
gene.annotations<-read.delim("/projects/jtopham_prj/Thesis_prj/data/mapping_files/gene_annotations.tsv",sep='\t',header=F,stringsAsFactors=F)

gene.annotations<-gene.annotations[!duplicated(gene.annotations[c("V2","V3")]),]
rownames(gene.annotations)<-NULL
colnames(gene.annotations)<-c("chromosome","ensembl_id","biotype")
# raw read counts generated by subread

featureCounts.data<-read.delim("/projects/jtopham_prj/Thesis_prj/results/HEK_RNAseq/HEK293_featCount_sept25.tsv",sep='\t',header=T,stringsAsFactors=F)
featureCounts.data<-featureCounts.data[,c(1,7,8,9,10)]
colnames(featureCounts.data)<-c("ENSMBL_ID","WT", "D320", "D351", "D372")


##------------------------------------------------------------------------------------------------------------------------
##  FUNCTIONS
##------------------------------------------------------------------------------------------------------------------------

## conduct bootstrapping procedure
tree.perm.test<-function(input.matrix, n){
  ## this function takes as input a matrix of normalized counts and integer 'n', specifying number of times to bootstrap
  ## output is vector of length 7, specifiying how many trees of each conformation were generated across n iterations
  perm.results<-rep(0,7)                 # 7 possible groupings
  for (i in 1:n){
    exp.subset<-input.matrix[sample(c(1:nrow(input.matrix)), size=nrow(input.matrix), replace = TRUE),]
    tree<-hclust(dist(t(exp.subset), method = "euclidean"), method = "ward.D" )
    cut_tree<-cutree(tree,k=2)           # cut tree into two groups
                                         # enumerate all 7 possibilities to 'label' this tree's structure
    if ( all(unname(cut_tree)==c(1,2,2,2)) | all(unname(cut_tree)==c(2,1,1,1)) ){perm.results[1]<-perm.results[1]+1}
    if ( all(unname(cut_tree)==c(2,1,2,2)) | all(unname(cut_tree)==c(1,2,1,1)) ){perm.results[2]<-perm.results[2]+1}
    if ( all(unname(cut_tree)==c(2,2,1,2)) | all(unname(cut_tree)==c(1,1,2,1)) ){perm.results[3]<-perm.results[3]+1}
    if ( all(unname(cut_tree)==c(2,2,2,1)) | all(unname(cut_tree)==c(1,1,1,2)) ){perm.results[4]<-perm.results[4]+1}
    if ( all(unname(cut_tree)==c(1,1,2,2)) | all(unname(cut_tree)==c(2,2,1,1)) ){perm.results[5]<-perm.results[5]+1}
    if ( all(unname(cut_tree)==c(1,2,1,2)) | all(unname(cut_tree)==c(2,1,2,1)) ){perm.results[6]<-perm.results[6]+1}
    if ( all(unname(cut_tree)==c(1,2,2,1)) | all(unname(cut_tree)==c(2,1,1,2)) ){perm.results[7]<-perm.results[7]+1}
  }
  return(perm.results)
}


##------------------------------------------------------------------------------------------------------------------------
##  SCRIPT
##------------------------------------------------------------------------------------------------------------------------

## DEA using DESeq2

# prepare for DESeq2: map ensembl IDs to gene names. note that some gene names map to multiple ensembl ids
rownames(featureCounts.data)<-featureCounts.data[,1]
rownames(featureCounts.data)<-paste(rownames(featureCounts.data),ensembl.mapper$gene_ID[match(rownames(featureCounts.data),ensembl.mapper$ensembl_ID)],sep='_')

# construct experimental design object
colData<-data.frame(cbind(condition=c("WT","KO","KO","KO"),type=c(rep("paired-end",4))),stringsAsFactors = F)
rownames(colData)<-colnames(featureCounts.data[,-1])
colData$condition<-factor(colData$condition,levels=unique(colData$condition))


if(file.exists('/projects/jtopham_prj/Thesis_prj/KMT2D_EZH2_paper/results/HEK_norm_counts.RData')){
  load('/projects/jtopham_prj/Thesis_prj/KMT2D_EZH2_paper/data/HEK_norm_counts_MLL2.RData')
  load('/projects/jtopham_prj/Thesis_prj/KMT2D_EZH2_paper/data/DSQ_object.RData')
}
if(!file.exists('/projects/jtopham_prj/Thesis_prj/KMT2D_EZH2_paper/results/')){
  # construct DESeq object
  DSq.Data<-DESeqDataSetFromMatrix(countData=featureCounts.data[,-1],colData=colData,design= ~condition)
  
  # perform DESeq2
  DSq.Object<-DESeq(DSq.Data)
  
  #extract normalized counts
  normalized.counts <- as.data.frame(counts(DSq.Object, normalized=TRUE))
  save(normalized.counts,file='/projects/jtopham_prj/Thesis_prj/KMT2D_EZH2_paper/data/HEK_norm_counts_MLL2.RData')
  save(DSq.Object,file='/projects/jtopham_prj/Thesis_prj/KMT2D_EZH2_paper/data/DSQ_object.RData')
}

# log transform normalized count data and filter for reads with expression greater than 10 in all samples
normalized.counts.log<-data.frame(apply(normalized.counts,2,FUN=function(x){x+1}),stringsAsFactors = F)
normalized.counts.log<-data.frame(apply(normalized.counts.log,2,log2),stringsAsFactors = F)
normalized.counts.log.filtered<-normalized.counts.log[unname(apply(normalized.counts.log,1,FUN=function(x){sum(x>=log2(10))==ncol(normalized.counts.log)})),]

if(file.exists('/projects/jtopham_prj/Thesis_prj/KMT2D_EZH2_paper/results/bootstrap.res.RData')){
  load('/projects/jtopham_prj/Thesis_prj/KMT2D_EZH2_paper/results/bootstrap.res.RData')}
if(!file.exists('/projects/jtopham_prj/Thesis_prj/KMT2D_EZH2_paper/results/bootstrap.res.RData')){
# construct different subsets of genes to use in clustering
test1<-as.matrix(normalized.counts.log.filtered) # test1 contains all genes (18,271)
rownames(test1)<-gsub('.*_','',rownames(test1))

test2<-test1[rownames(test1)%in%cosmic.genes[,1],] # test2 contains only cosmic cancer census genes (502)
test3<-test1[rownames(test1)%in%hg19.genes[hg19.genes[,1]=="12",2],] # test3 contains only genes on chr 12 (956)
test4<-test1[rownames(test1)%in%c(ortega),] # test4 contains DE genes (downreg.) from Ortega paper; mouse adjusted p < 0.05 (250)

# bootstrap 100k times with the original and 3 filtered sets
btsp.unfiltered<-tree.perm.test(test1,100000)
btsp.filtered<-tree.perm.test(test2,100000)
btsp.filtered12<-tree.perm.test(test3,100000)
btsp.filtered.ortega<-tree.perm.test(test4,100000)

# combine results into single dataframe
btsp.res<-data.frame(rbind(cbind(Dataset="All genes",Structure=c(1:7),Count=btsp.unfiltered),
                           cbind(Dataset="COSMIC",Structure=c(1:7),Count=btsp.filtered),
                           cbind(Dataset="Chr 12",Structure=c(1:7),Count=btsp.filtered12),
                           cbind(Dataset="Ortega-Molina 2016",Structure=c(1:7),Count=btsp.filtered.ortega)),
                     stringsAsFactors=F)
btsp.res$Count<-(as.numeric(btsp.res$Count)/100000)
save(btsp.res,file='/projects/jtopham_prj/Thesis_prj/KMT2D_EZH2_paper/results/bootstrap.res.RData')
}

# plot distribution of tree structures across the 4 sets
p<-ggplot(btsp.res, aes(x=Dataset,y=Count,fill=Structure))
p+geom_bar(stat="identity",alpha=0.8,position="stack")+
  theme(axis.text.y=element_text(size = 32, colour = "black",face="bold"),
        axis.text.x=element_text(size = 32, colour = "black",face="bold",angle=50,hjust=1),
        legend.title=element_text(size=29),legend.text=element_text(size=24),
        panel.background = element_rect(fill="white"),
        axis.title.y=element_text(size=37,face="bold",colour="black",margin=margin(0,33,0,0)))+
  ylab("Proportion")+xlab("")+ggtitle("")+
  scale_y_continuous(breaks=seq(0, 1, .2))+
  scale_fill_manual(guide=guide_legend(title=""),values=c(pair_pal[c(2,1,3,5,7,4)],"purple"))

# show heatmap for the Ortega-filtered dataset
annotation <- data.frame(Var1 = factor(c("WT","KO","KO","KO"), levels = c("WT","KO")))
rownames(annotation) <- colnames(test4)
ann.colors = list(Var1 = c(WT=pair_pal[3],KO=pair_pal[4]))
pheatmap(test4,show_rownames = F,show_colnames=F,color=blue_pal,annotation=annotation,annotation_colors=ann.colors)


## Parse and visualize DEA results
RNAseq.results<-data.frame(results(DSq.Object),stringsAsFactors = F)
RNAseq.results<-RNAseq.results[order(RNAseq.results$padj),]

RNAseq.results<-cbind(RNAseq.results,neg.log10.padj=-log10(RNAseq.results$padj))  # add column for -log10 tranformed adjusted p values
RNAseq.results<-cbind(RNAseq.results,abs.FC=abs(RNAseq.results$log2FoldChange))   # add column for absolute val fold change
RNAseq.results<-cbind(RNAseq.results,gene.id=gsub('.*_','',rownames(RNAseq.results)))  # add column for gene id

RNAseq.results<-cbind(RNAseq.results,log2baseMean=log2(RNAseq.results$baseMean))  # add column for log2 baseMean
RNAseq.results<-cbind(RNAseq.results,pval.threshold=RNAseq.results$padj<0.05)     # add column for selective coloring
RNAseq.results$pval.threshold[is.na(RNAseq.results$pval.threshold)]<-"FALSE"
RNAseq.results$pval.threshold[RNAseq.results$pval.threshold=="FALSE"]<-"Adj. p-value > 0.05"
RNAseq.results$pval.threshold[RNAseq.results$pval.threshold=="TRUE"]<-"Adj. p-value < 0.05"

RNAseq.results$pval.threshold<-factor(RNAseq.results$pval.threshold,levels=c("Adj. p-value < 0.05","Adj. p-value > 0.05"))

# visualize DEA results
p<-ggplot(RNAseq.results,aes(x=log2baseMean))
p1<-p+geom_density(fill=blue_pal[3],alpha=0.9)+ xlab("log(2) Average Expression")+
  theme(axis.text.x = element_text(size=32,color="black",face="bold"),axis.text.y=element_text(size=32,color="black",face="bold"),
        legend.text = element_text(size = 23, colour = "black"),legend.title=element_text(size=25),
        panel.background = element_rect(fill="white"),panel.grid.major = element_line(linetype="dashed",colour =ygob_pal[233]),
        panel.grid.minor = element_line(linetype="dashed",colour = ygob_pal[233]),
        axis.title=element_text(size=38,color="black",face="bold"))

p<-ggplot(RNAseq.results,aes(x=log2baseMean,y=log2FoldChange))
p2<-p+geom_point(size=6.5,aes(color=pval.threshold),alpha=0.7)+
  xlab("log(2) Average Expression")+ylab("log(2) Fold Change")+
  theme(axis.text.x = element_text(size=52,color="black",face="bold"),axis.text.y=element_text(size=52,color="black",face="bold"),legend.background = element_rect(fill=NA, size=.5),
        legend.text = element_text(size = 48,face="bold",colour = "black"),legend.position=c(.89,.89),
        axis.title.y=element_text(size=68,color="black",face="bold",margin = margin(0,32,0,0)),legend.key.width=unit(2,"line"),legend.key.size=unit(2,"line"),
        axis.title.x=element_text(size=68,color="black",face="bold",margin = margin(32,0,0,0)),
        legend.key=element_rect(fill="white",color="black"),plot.margin=unit(c(1,1.5,1,1),"cm"),
        panel.background = element_rect(fill="white"),panel.grid.major = element_line(linetype="dashed",colour =ygob_pal[233]),
        panel.grid.minor = element_line(linetype="dashed",colour = ygob_pal[233]))+
  scale_color_manual(guide = guide_legend(title = "",override.aes = list(size=22)),values=c(blue_pal[12],blue_pal[2]))
#+geom_text(data=subset(RNAseq.results[RNAseq.results$gene.id%in%c("CSPG4","COL5A1","COL9A3","S100A16"),], log2FoldChange > 1.25 | log2FoldChange < -2.2), aes(log2baseMean,log2FoldChange,label=gene.id),
#vjust = -.6,color="black",size=7)

# stack two plots vertically using gridExtra
gA <- ggplotGrob(p2)
gB <- ggplotGrob(p1)
maxWidth = grid::unit.pmax(gA$widths[2:5], gB$widths[2:5])
gA$widths[2:5] <- as.list(maxWidth)
gB$widths[2:5] <- as.list(maxWidth)
grid.arrange(gA, gB, ncol=1,heights=c(1, 0.4))

RNAseq.results<-na.omit(RNAseq.results)
save(RNAseq.results,file="/projects/jtopham_prj/Thesis_prj/KMT2D_EZH2_paper/results/HEK_DEA_results.RData")
